### ISSUE TITLE
After upgrade from 1.3.1 to 1.4.0 get_storage  i get type 'LastCall' is not a subtype of type 'Map<String, dynamic>' error

### ISSUE THREAD

**Issue by halkportal970** [jonataslaw/get_storage#41]
After 1.4.0 upgrade,  I get 
**type 'LastCall' is not a subtype of type 'Map<String, dynamic>'**
error.
I have a model called LastCall, and I store data as json .
First call of loadLastCallList() function works perfect. Bu second time, third time etc fails.
First time  **list** variable is  a  **[0]: Map**
But the other times **list** variable is a **[0]:LastCall**
for that  **fromJson** fails to convert **list[i]**

Why **list** variable is first time  is a Map but the other times  is a Model(LastCall)
 **List<dynamic> list = GetStorage().read("lastCallList");**
I am excepting it all the time as a Map variable.

```
Future<void> loadLastCallList() async {
    var tempList = List<LastCall>();
      List<dynamic> list = GetStorage().read("lastCallList");
      if (list != null) {
        for (var i = 0; i < list.length; i++) {
          LastCall lastCall = LastCall.fromJson(list[i]);
          tempList.add(lastCall);
        }
      } else {
        lastCallList.clear();
      }
      lastCallList.assignAll(tempList);
}
```

**Comment by halkportal970** [comment]
After I downgrade to 1.3.2 code works fine.

**Comment by jpandco** [comment]
yes the same error. if you write Model then read he will return from memory and the return will keep the Type Model.
but if the model was not in memory he will get the model from storage and the return will be  _InternalLinkedHashMap<String, dynamic>

**Comment by jpandco** [comment]
> After 1.4.0 upgrade, I get
> **type 'LastCall' is not a subtype of type 'Map<String, dynamic>'**
> error.
> I have a model called LastCall, and I store data as json .
> First call of loadLastCallList() function works perfect. Bu second time, third time etc fails.
> First time **list** variable is a **[0]: Map**
> But the other times **list** variable is a **[0]:LastCall**
> for that **fromJson** fails to convert **list[i]**
> 
> Why **list** variable is first time is a Map but the other times is a Model(LastCall)
> **List list = GetStorage().read("lastCallList");**
> I am excepting it all the time as a Map variable.
> 
> ```
> Future<void> loadLastCallList() async {
>     var tempList = List<LastCall>();
>       List<dynamic> list = GetStorage().read("lastCallList");
>       if (list != null) {
>         for (var i = 0; i < list.length; i++) {
>           LastCall lastCall = LastCall.fromJson(list[i]);
>           tempList.add(lastCall);
>         }
>       } else {
>         lastCallList.clear();
>       }
>       lastCallList.assignAll(tempList);
> }
> ```

i think this will fix your error

  Future<void> loadLastCallList() async {
    var tempList = List<LastCall>();
    var _data = GetStorage().read("lastCallList");
    
    var list = _data is List<LastCall> ? _data : Map<String, dynamic>.from(_data).entries.toList();
    
    if (list != null) {
      for (var i = 0; i < list.length; i++) {
        var _lastCall = list[i] is LastCall ? list[i] : LastCall.fromJson(list[i]);
        tempList.add(_lastCall);
      }
    } else {
      lastCallList.clear();
    }
    lastCallList.assignAll(tempList);
  }

**Comment by AlirezaDaryani** [comment]
same problem, i build model data with fromJson and toJson, but 'read' dose not give me my model.

**Comment by dreamer2q** [comment]
When your `data` is stored in memory, you can get the exact type as you want, otherwise you will get json decoded data from LocalStorage (usually typed `Map` or `List` depending your data type).

Therefore, you have to check the data type before you try parse `fromJson`.

This is the problem of flutter, you have to compromise.

**Comment by iammazharul** [comment]
> {

may be i am late, but i think it's still helpful. because me also faced the same problem. in my case i am forget to convert it in JSON format before wright action.In my case:

error: get_storage is not a subtype of type 'Map<String, dynamic>'
wrong way:          
           box.write('user', data.userData);

right way:
          UserData userData;
          GetStorage box = GetStorage();
          box.write('user', data.userData.toJson());
          
          userData = UserData.fromJson(box.read('user'));

**Comment by aminkhan1** [comment]
i had this problem. i fix this problem like this
`   
 if (store.read('docs') is DocsModel) {
      return model = store.read('docs');
    } else {
      return model = DocsModel.fromJson(store.read('docs'));
    }
`

**Comment by hhstore** [comment]
> may be i am late, but i think it's still helpful. because me also faced the same problem. in my case i am forget to convert it in JSON format before wright action.In my case:
> 
> error: get_storage is not a subtype of type 'Map<String, dynamic>'
> wrong way:
> ` box.write('user', data.userData);`
> 
> right way:
> 
> ```
>           UserData userData;
>           GetStorage box = GetStorage();
>           box.write('user', data.userData.toJson());
> ```
> 
> ` userData = UserData.fromJson(box.read('user'));`


## Hive DB: 

- like HiveDB: https://github.com/hivedb/hive#store-objects
- use `hive_generator` automatically generate `TypeAdapters` for almost any class.
-  register a `TypeAdapter` which converts the object `from and to` binary form.
- https://docs.hivedb.dev/#/custom-objects/generate_adapter

![image](https://user-images.githubusercontent.com/3252130/126429258-44a7dda4-7843-4dde-9dab-0fe153f4bb3e.png)


## Protobuf:

- The data definition for `hive`, very similar to the way `protobuf` is defined. 
- In fact, `get storage` can directly support `protobuf` to define the data model. Wouldn't it be better?
- https://github.com/dart-lang/protobuf

**Comment by hhstore** [comment]
# ref: 

- https://github.com/dart-lang/sdk/issues/46854



## solution: 

- for dart `null safety`. 
- fix here: 

```dart



  /// get many:
  /// model 需要实现 fromJson(), toJson() 方法
  Map<String, UserEntity> users() {
    var key = keys.users();

    /// (UserEntity, has fromJson, toJson)
    var raw = kv.read(key) ?? {};
    print('users:$raw, ${raw.runtimeType}');

    // if (raw is Map<String, UserEntity>) {
    //   return raw;
    // }

    // return raw;
    var result = <String, UserEntity>{};

    raw.forEach((key, value) {
      print('$key, $value, ${value.runtimeType}');
      result[key] = value is UserEntity ? value : UserEntity.fromJson(value); // TODO: here!!!
      // result[key] = (value as UserEntity?)!;
    });

    return result;
    // return raw ?? {};
    // return Map<String, UserEntity>.from(out ?? {});
  }

```
